keps/sig-api-machinery/20190721-explain-categories-expand.md

```diff
@@ -0,0 +1,198 @@
+```
+
+The value (here *1.category-name*) contains two parts, separated by a dot (.).
+The first part is an integer that indicates the order of the category related to the
+other categories of the field. The second part is a name of the category, with spaces
+replaced by dashes (-).
+
+Example:
+```go
+type PodSpec struct {
+  // +k8s:openapi-gen=x-kubernetes-explain-category:2.volumes
+  Volumes []Volume
+  // +k8s:openapi-gen=x-kubernetes-explain-category:1.containers
+  InitContainers []Container
+  // +k8s:openapi-gen=x-kubernetes-explain-category:1.containers
+  Containers []Container
+  // +k8s:openapi-gen=x-kubernetes-explain-category:4.lifecycle
+  RestartPolicy RestartPolicy
+  [...]
+```
+
+As a result, the documentation of PodSpec would be written by `kubectl explain` as:
+```
+KIND:     Pod
+VERSION:  v1
+
+RESOURCE: spec <Object>
+
+DESCRIPTION:
+     ...
+     
+FIELDS:
+  CONTAINERS
+
+   initContainers        <[]Object>
+     ...
+
+   containers            <[]Object>
+     ...
+
+  VOLUMES
+
+   volumes               <[]Object>
+     ...
+
+  LIFECYCLE
+  
+   restartPolicy         <Object>
+    ...
+```
+
+#### Expand some fields of Container
+
+To indicate that a field should be expanded in the documentation, it is necessary to add a comment to the field in `types.go`:
+
+```
+// +k8s:openapi-gen=x-kubernetes-explain-expand:
+```
+
+Example:
+
+```go
+type Container struct {
+  // +k8s:openapi-gen=x-kubernetes-explain-expand:
+  Ports []ContainerPort
+  [...]
+}
+```
+
+As a result, the documentation of Container would be written by `kubectl explain` as:
+
+```
+KIND:     Pod
+VERSION:  v1
+
+RESOURCE: containers <[]Object>
+
+DESCRIPTION:
+     ...
+
+FIELDS:
+
+   [first Container fields...]
+
+   ports     <[]Object>
+     <description of ports>
+
+     hostIP     <string>
+       <description of hostIP>
+
+     hostPort   <sint32>
+       <description of hostPort>
+
+     [other ports fields]
+
+   [other Container fields...]
+```
+
+#### Order fields of VolumeMount
+
+In order to sort fields without indicating a category, it is necessary to add a comment to the field in `types.go`:
+
+```
+// +k8s:openapi-gen=x-kubernetes-explain-category:1.
+```
+
+The value (here *1.*) should be an integer followed by a dot (.), indicating 
+the order of this field related to the other fields in the same Kind.
+
+Example:
+```go
+type VolumeMount struct {
+  // +k8s:openapi-gen=x-kubernetes-explain-category:1.
+  Name string
+  // +k8s:openapi-gen=x-kubernetes-explain-category:2.
+  ReadOnly bool 
+  // +k8s:openapi-gen=x-kubernetes-explain-category:2.
+  MountPath string
+  // +k8s:openapi-gen=x-kubernetes-explain-category:3.
+  SubPath string 
+}  
+```
+
+Without annotations, the fields would be displayed in the following order (alphabetically):
+- mountPath
+- name
+- readOnly
+- subPath
+
+With the annotations, the fields can now be displayed in this order:
+- name
+- mountPath
+- readOnly
+- subPath
+
+## Compatibility
```

Comment:
I would like to see this proposal discuss how categorization fits in with existing tabular output. They seem highly related, like maybe this categorization should drive the column headers or something. Also, I think something controlling "visible by default" vs "visible with -o wide" should be delivered by something like this; I'd like to see the proposal explain how these things work together (or don't).